sympy/functions/elementary/exponential.py

```diff
@@ -1007,15 +1009,22 @@ def coeff_exp(term, x):
             while s.is_Order:
                 n += 1
                 s = arg.nseries(x, n=n, logx=logx)
-        a, b = s.removeO().leadterm(x)
+        if logx.has(x) and logx != log(x):
```

Comment:
Is the `and` clause really required? When is `logx` = `log(x)`?

Reply:
Not really , I would say no , actually that shoudn't be there , put it just by choice . From what I understand the parameter `logx` should be replacing all instances of `log(x)` in the series/leading term. For eg
```
>>> (log(x) + log(1+x)).nseries(x,n=3, logx = p)
x - x**2/2 + p + O(x**3)
```
**But quite some times we encounter `log(x)` as the `logx` parameter, which I guess means there shouldn't be any change in the expression correct ?( like replacing log(x) by log(x) )**
```
 >>> (log(x) + log(1+x)).nseries(x,n=3, logx = log(x))
log(x) + x - x**2/2 + O(x**3)
```
So yeah few tests like the following involve such a change and I had put that check by choice so that the output looks somewhat better
```
>>> log(1 + 1/x).nseries(x, n=5) # calls for log(1 + 1/x)._eval_nseries(x, n=5 , logx=log(x))
log(1/x) + x - x**2/2 + x**3/3 - x**4/4 + O(x**5)    # without the check

x - log(x) - x**2/2 + x**3/3 - x**4/4 + O(x**5)      # with the check 
```
if log(x) replaces log(x) it means there shouldn't be any change from the normal series expansion so the answer should be 
```
>>> log(1 + 1/x).series(x, 0, 5)
log(1/x) + x - x**2/2 + x**3/3 - x**4/4 + O(x**5)
```
Also another case being
```
>>> e = -log(x) + x*(-log(x) + log(sin(2*x))) + log(sin(2*x))
>>> e.series(x, n=5)
log(2*x) - log(x) + x*log(2) - 2*x**2/3 - 2*x**3/3 - 4*x**4/45 + O(x**5)
>>> # The first -log(x) doesn't really take part in the expansion but with the check we get
log(2) + log(2)*x - 2*x**2/3 - 2*x**3/3 - 4*x**4/45 + O(x**5)
```
So yeah , we could do with or without the check !

Reply:
The check looks counterintuitive to me, so I think it can be removed. I meant to ask if there are any example cases where we don't explicitly pass `logx=log(x)`, but it gets passed during the run of the code.

Reply:
Ahh ok got your point , will look into this . If I remember correctly , there are cases in `test_aseries.py` and `test_nseries.py`, where somehow a `log(x)` is being passed as the parameter. Not sure of its usecase and like where is it happening but I'll look into this soon

Reply:
> The check looks counterintuitive to me, so I think it can be removed. I meant to ask if there are any example cases where we don't explicitly pass logx=log(x), but it gets passed during the run of the code.

Actually I realize we would need some kind of check here like `logx != log(x)`. This is because , I had overlooked line `976-977` . 
```
    def _eval_nseries(self, x, n, logx, cdir=0):
        from sympy.series.order import Order
        from sympy.simplify.simplify import logcombine
        if not logx:
            logx = log(x)
```
So whenever we pass `logx` as `None` originally , it is changed to `log(x)` and the check I've added is for addressing a corner case as follows
```
def test_issue_21227():
    f = log(x)
    assert f.nseries(x, logx=-x) == -x

    f = log(-log(x))
    assert f.nseries(x, logx=-x) == log(x)

    f = log(log(x))
    assert f.nseries(x, logx=-x) == log(-x)
    assert f.nseries(x, logx=x) == log(x)
```
Here `logx` contains an argument based on `x` which is the variable in consideration for the expansion for which there is no handling in the code currently ! The original code is capable to address all other cases through `a, b = s.removeO().leadterm(x)` . We should have the following diff I believe
```
--- a/sympy/functions/elementary/exponential.py
+++ b/sympy/functions/elementary/exponential.py
@@ -973,6 +973,7 @@ def _eval_nseries(self, x, n, logx, cdir=0):
         #      IMPORTANT.
         from sympy.series.order import Order
         from sympy.simplify.simplify import logcombine
+        _logx = logx
         if not logx:
             logx = log(x)
         if self.args[0] == x:
@@ -1009,10 +1010,13 @@ def coeff_exp(term, x):
             while s.is_Order:
                 n += 1
                 s = arg.nseries(x, n=n, logx=logx)
-        if logx.has(x):
+        if _logx and logx.has(x):
             a, b = arg.subs(log(x), logx).as_leading_term(x), S.Zero
         else:
-            a, b = s.removeO().leadterm(x)
+            try:
+                a, b = s.removeO().leadterm(x)
+            except:
+                a, b = s.removeO().as_leading_term(x), S.Zero
```
The last try except block would be for something like this
```
>>> log(sqrt(-x**2 -2*x))._eval_nseries(x, 3, None)
log(sqrt(2)*sqrt(-x)) + x/4 - x**2/16 + O(x**3)
>>> 
>>> # Without the block 
>>> log(sqrt(-x**2 -2*x))._eval_nseries(x, 3, None)"
ValueError:
cannot compute leadterm(-sqrt(2)*x**2*sqrt(-x)/32 +
sqrt(2)*x*sqrt(-x)/4 + sqrt(2)*sqrt(-x), x). The coefficient should
have been free of x but got sqrt(2)*sqrt(-x)
```